src/libsyntax/sess.rs

```diff
@@ -13,45 +13,85 @@ use syntax_pos::{Symbol, Span, MultiSpan};
-    /// Spans collected for gating `crate_visibility_modifier`, e.g. `crate fn`.
-    pub crate_visibility_modifier: Lock<Vec<Span>>,
-    /// Spans collected for gating `const_generics`, e.g. `const N: usize`.
-    pub const_generics: Lock<Vec<Span>>,
-    /// Spans collected for gating `decl_macro`, e.g. `macro m() {}`.
-    pub decl_macro: Lock<Vec<Span>>,
-    /// Spans collected for gating `box_patterns`, e.g. `box 0`.
-    pub box_patterns: Lock<Vec<Span>>,
-    /// Spans collected for gating `exclusive_range_pattern`, e.g. `0..2`.
-    pub exclusive_range_pattern: Lock<Vec<Span>>,
-    /// Spans collected for gating `try_blocks`, e.g. `try { a? + b? }`.
-    pub try_blocks: Lock<Vec<Span>>,
-    /// Spans collected for gating `label_break_value`, e.g. `'label: { ... }`.
-    pub label_break_value: Lock<Vec<Span>>,
-    /// Spans collected for gating `box_syntax`, e.g. `box $expr`.
-    pub box_syntax: Lock<Vec<Span>>,
-    /// Spans collected for gating `type_ascription`, e.g. `42: usize`.
-    pub type_ascription: Lock<Vec<Span>>,
+pub struct GatedSpans {
+    crate spans: Lock<FxHashMap<Symbol, Vec<Span>>>,
+}
+
+impl GatedSpans {
+    /// Feature gate the given `span` under the given `feature`
+    /// which is same `Symbol` used in `active.rs`.
+    pub fn gate(&self, feature: Symbol, span: Span) {
+        self.spans
+            .borrow_mut()
+            .entry(feature)
+            .or_default()
+            .push(span);
+    }
+
+    /// Ungate the last span under the given `feature`.
+    /// Panics if the given `span` wasn't the last one.
+    ///
+    /// Using this is discouraged unless you have a really good reason to.
+    pub fn ungate_last(&self, feature: Symbol, span: Span) {
+        let removed_span = self.spans
+            .borrow_mut()
+            .entry(feature)
+            .or_default()
+            .pop()
+            .unwrap();
+        debug_assert_eq!(span, removed_span);
+    }
+
+    /// Is the provided `feature` gate ungated currently?
+    ///
+    /// Using this is discouraged unless you have a really good reason to.
+    pub fn is_ungated(&self, feature: Symbol) -> bool {
+        self.spans
+            .borrow()
+            .get(&feature)
+            .map(|spans| spans.is_empty())
+            .unwrap_or(true)
+    }
+
+    /// Take a snapshot of the state of pre-expansion gating at this point.
+    /// Once we are ready to `snapshot.commit();` the spans at this point
+    /// and the ones that have since been added will be combined.
+    pub fn snaphot(&self) -> GatedSpansSnapshot<'_> {
+        GatedSpansSnapshot {
+            gated_spans: self,
+            existing: mem::take(&mut self.spans.borrow_mut()),
+        }
+    }
+}
+
+/// A snapshot of the existing `GatedSpans`.
+pub struct GatedSpansSnapshot<'a> {
+    gated_spans: &'a GatedSpans,
+    /// These are the spans that were gated before taking the snaphot.
+    existing: FxHashMap<Symbol, Vec<Span>>,
+}
```

Comment:
I don't think this needs a separate struct?
Note that if you pass `gated_spans` as an argument to commit/rollback, the only thing that will leave here is equivalent to `GatedSpans` itself.
So I'd rather add a single `merge` method to `GatedSpans` and use `mem::take` or `mem::swap` directly otherwise.

Reply:
On the other hand, it looks like rollback (or commit) is better packed into a destructor, RAII style, then we'll need the struct anyway.

Reply:
Tricky... I decided to go with the first approach. Wdyt?